bluepysnap/circuit_ids.py

```diff
@@ -0,0 +1,222 @@
+            data extraction from a sorted dataframe.
+        """
+        if not inplace:
+            return CircuitNodeIds(self.index, sort_index=True)
+        self.index = self.index.sortlevel()[0]
+        return None
+
+    def append(self, other, inplace=False):
+        """Append a CircuitNodeIds to the current one.
+
+        Args:
+            other (CircuitNodeIds): the other CircuitNodeIds to append to the current one.
+            inplace (bool): if set to True. Do the transformation inplace.
+        """
+        if not inplace:
+            return CircuitNodeIds(self.index.append(other.index))
+        self.index = self.index.append(other.index)
+        return None
+
+    def sample(self, sample_size, inplace=False):
+        """Sample a CircuitNodeIds.
+
+        Notes:
+            this function takes randomly ``sample_size`` values of a circuit node ids.
+
+        Args:
+            sample_size (int): the size of the sample. If the size of the sample is greater than
+                the size of the CircuitNodeIds then all ids are taken and shuffled.
+            inplace (bool): if set to True. Do the transformation inplace.
+        """
+        res = self if inplace else self.copy()
+        if len(res) != 0:
+            sample_size = sample_size if sample_size < len(res) else len(res)
+            res.index = res.index[np.random.choice(len(res), size=sample_size, replace=False)]
+        if not inplace:
+            return res
+        return None
+
+    def limit(self, limit_size, inplace=False):
+        """Limit the size of a CircuitNodeIds.
+
+        Notes:
+            this function takes the first ``limit_size`` values of a circuit node ids.
+
+        Args:
+            limit_size (int): the size of the sample. If the size of the sample is greater than
+                the size of the CircuitNodeIds then all ids are kept.
+            inplace (bool): if set to True. Do the transformation inplace.
+        """
+        res = self if inplace else self.copy()
+        res.index = res.index[0: limit_size]
+        if not inplace:
+            return res
+        return None
+
+    def to_csv(self, filepath):
+        """Save CircuitNodeIds to csv format."""
+        self.index.to_frame(index=False).to_csv(filepath, index=False)
+
+    @classmethod
+    def from_csv(cls, filepath):
+        """Load CircuitNodeIds from csv."""
+        return cls(pd.MultiIndex.from_frame(pd.read_csv(filepath)))
+
+    def __repr__(self):
+        """Correct repr of the CircuitNodeIds."""
+        res = self.index.__repr__()[len("MultiIndex"):]
+        return "CircuitNodeIds" + res
+
+    def __str__(self):
+        """Correct str of the CircuitNodeIds."""
+        return self.__repr__()
+
+    def __eq__(self, other):
+        """Equality for the CircuitNodeIds."""
+        if not isinstance(other, CircuitNodeIds):
+            return False
+        return self.index.equals(other.index)
+
+    def __len__(self):
```

Comment:
Have you tried implementing directly \_\_iter\_\_ instead of \_\_len\_\_ ?

Reply:
Maybe this comment was not clear.
I *suppose* that if you do:

```python
def __iter__(self):
    return iter(self.index)
```
then len(self) should be len(self.index) automatically ?

Reply:
I have a \_\_iter\_\_ and a \_\_len\_\_ now.